design/backup-resource-list.md

```diff
@@ -0,0 +1,96 @@
+# Expose list of backed up resources in backup details
+
+Status: Draft
+
+To increase the visibility of what a backup might contain, this document proposes storing metadata about backed up resources in object storage and adding a new section to the detailed backup description output to list them.
+
+## Goals
+
+- Include a list of backed up resources as metadata in the bucket
+- Enable users to get a view of what resources are included in a backup using the Velero CLI
+
+## Non Goals
+
+- Expose the full manifests of the backed up resources
+
+## Background
+
+As reported in [#396](https://github.com/heptio/velero/issues/396), the information reported in a `velero backup describe <name> --details` command is fairly limited, and does not easily describe what resources a backup contains.
+In order to see what a backup might contain, a user would have to download the backup tarball and extract it.
+This makes it difficult to keep track of different backups in a cluster.
+
+## High-Level Design
+
+After performing a backup, a new file will be created that contains the list of the resources that have been included in the backup.
+This file will be persisted in object storage alongside the backup contents and existing metadata.
+
+A section will be added to the output of `velero backup describe <name> --details` command to view this metadata.
+
+## Detailed Design
+
+### Metadata file
+
+This metadata will be in JSON (or YAML) format so that it can be easily inspected from the bucket outside of Velero tooling, and will contain the API resource and group, namespaces and names of the resources:
+
+```
+apps/v1/Deployment:
+- default/database
+- default/wordpress
+v1/Service:
+- default/database
+- default/wordpress
+v1/Secret:
+- default/database-root-password
+- default/database-user-password
+v1/ConfigMap:
+- default/database
+v1/PersistentVolume:
+- my-pv
+```
+
+The filename for this metadata will be `<backup name>-resource-list.json`.
+The top-level key is the string form of the `schema.GroupResource` type that we currently keep track of in the backup controller code path.
+
+### Changes in Backup controller
+
+The Backupper currently initialises a map to track the `backedUpItems` (https://github.com/heptio/velero/blob/1594bdc8d0132f548e18ffcc1db8c4cd2b042726/pkg/backup/backup.go#L269), this is passed down through GroupBackupper, ResourceBackupper and ItemBackupper where ItemBackupper records each backed up item.
+This property will be moved to the [Backup request struct](https://github.com/heptio/velero/blob/16910a6215cbd8f0bde385dba9879629ebcbcc28/pkg/backup/request.go#L11), allowing the BackupController to access it after a successful backup.
+
+`backedUpItems` currently uses the `schema.GroupResource` as a key for the resource.
```

Comment:
Is there a function that allows for easy parsing of this format back into a GVR?

Reply:
Do you mean converting from `schema.GroupResource` to `schema.GroupVersionKind` or the other way around? We can't get the GVK from the GroupResource, but we can get it from the `runtime.Unstructured` obj. We don't really need the GroupResource, AFAICT we're only using it as a key in backedUpItems and not using it for anything else after.

Reply:
Sorry, should have commented lower down. I meant the `{group}/{version}/{kind}` string, and I meant GVK, not GVR. I suppose writing a deserialization function would be simple enough.

It may not actually be all that important, as we could just print the key out verbatim when displaying details.

Reply:
Ah I see what you mean now, I was just thinking of printing out the key as-is, but we wouldn't be able to directly parse that string (since the GVK.String() method has a slightly different format), but (without looking at the schema pkg) I believe we can parse the `schema.GroupVersion` using `{group}/{version}` and then convert that to GVK using the remaining part.